hedera-node/src/main/java/com/hedera/services/ledger/interceptors/EndOfStakingPeriodCalculator.java

```diff
@@ -0,0 +1,149 @@
+import com.hedera.services.utils.EntityNum;
+import com.hederahashgraph.api.proto.java.NodeStake;
+import com.hederahashgraph.api.proto.java.Timestamp;
+import com.swirlds.merkle.map.MerkleMap;
+
+import javax.inject.Inject;
+import javax.inject.Singleton;
+import java.time.Instant;
+import java.util.ArrayList;
+import java.util.List;
+import java.util.function.Supplier;
+
+import static com.hedera.services.records.TxnAwareRecordsHistorian.DEFAULT_SOURCE_ID;
+import static com.hedera.services.state.EntityCreator.NO_CUSTOM_FEES;
+
+@Singleton
+public class EndOfStakingPeriodCalculator {
+	public static final String END_OF_STAKING_PERIOD_CALCULATIONS_MEMO = "End of Staking Period Calculation record";
+
+	private final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts;
+	private final Supplier<MerkleMap<EntityNum, MerkleStakingInfo>> stakingInfoSupplier;
+	private final Supplier<MerkleNetworkContext> merkleNetworkContextSupplier;
+	private final SyntheticTxnFactory syntheticTxnFactory;
+	private final RecordsHistorian recordsHistorian;
+	private final EntityCreator creator;
+	private final PropertySource properties;
+
+	@Inject
+	public EndOfStakingPeriodCalculator(
+			final Supplier<MerkleMap<EntityNum, MerkleAccount>> accounts,
+			final Supplier<MerkleMap<EntityNum, MerkleStakingInfo>> stakingInfoSupplier,
+			final Supplier<MerkleNetworkContext> merkleNetworkContextSupplier,
+			final SyntheticTxnFactory syntheticTxnFactory,
+			final RecordsHistorian recordsHistorian,
+			final EntityCreator creator,
+			@CompositeProps PropertySource properties
+	) {
+		this.accounts = accounts;
+		this.stakingInfoSupplier = stakingInfoSupplier;
+		this.merkleNetworkContextSupplier = merkleNetworkContextSupplier;
+		this.syntheticTxnFactory = syntheticTxnFactory;
+		this.recordsHistorian = recordsHistorian;
+		this.creator = creator;
+		this.properties = properties;
+	}
+
+	public void updateNodes(final Instant consensusTime) {
+		final var stakingInfo = stakingInfoSupplier.get();
+		final var merkleNetworkContext = merkleNetworkContextSupplier.get();
+
+		// skip end of staking period calculations if the rewards are not yet activated.
+		if (!merkleNetworkContext.areRewardsActivated()) {
+			return;
+		}
+
+		final var rewardRate = Math.min(stakingRewardsAccountBalance(),
+				properties.getLongProperty("staking.rewardRate"));
+
+		long updatedTotalStakedRewardStart = 0L;
+		long updatedTotalStakedStart = 0L;
+		List<NodeStake> nodeStakingInfos = new ArrayList<>();
+
+		for (final var nodeNum : stakingInfo.keySet()) {
+			final var merkleStakingInfo = stakingInfo.getForModify(nodeNum);
+			merkleStakingInfo.updateRewardSumHistory(rewardRate, merkleNetworkContext.getTotalStakedRewardStart());
+
+			final var totalStake = merkleStakingInfo.getStakeToReward() + merkleStakingInfo.getStakeToNotReward();
+			final var minStake = merkleStakingInfo.getMinStake();
+			final var maxStake = merkleStakingInfo.getMaxStake();
+
+			if (totalStake > maxStake) {
```

Comment:
‼️  Do we really want to change state here? I would have expected we just call a method like `merkleStakingInfo.getEffectiveStake()` whose implementation is like,
```
public long getEffectiveStake() {
    return Math.min(maxStake, Math.max(Long.signum(stake - minStake), 0) * stake);
}
```

Reply:
We are supposed to change the state. At least that is what I understood from the HIP
```
// total tinybars of reward earned by all stakers for the staking period now ending
    long rate = min(accountBalance(0.0.800), stakingRewardRate); 
    
    for each node {
        for (i=365; i>0; i--) //push a day onto the history (latest reward is zero, but updated below)
            node.rewardSumHistory[i] = node.rewardSumHistory[i-1];
        //if this node was "active", then it should give rewards for this staking period
        if (node.numRoundsWithJudge / numRoundsInPeriod >= activeThreshold) 
            node.rewardSumHistory[0] += 
                rate * node.stakedRewardedStart / totalStakedRewardedStart / 100_000_000;
    }
    totalStakedRewardStart = 0;
    totalStakedNoRewardStart = 0;
    for each node {
        node.stake = node.stakedRewarded + node.stakedNotRewarded;
        if (node.stake > node.maxStake) node.stake = node.maxStake;   // <------------------------ here
        if (node.stake < node.minStake) node.stake = 0;                            // <------------------------ here
        long stakedRewardedUsed = min(node.stakedRewarded, node.stake);
        totalStakedRewardedStart += stakedRewardedUsed;
        totalStakedStart += node.stake;
    }              
    ```

Reply:
Surely this is just for the purposes of calculation, right? 

Otherwise this one step would erase all the work we did in `handleTransaction` during that day to track staking changes!

For example, at midnight of a day, `node0` has zero stake and `minStake = 1M`. During the day,
 - Alice stakes 250k tinybar.
 - Bob stakes 100k tinybar.
At midnight, we now "forget" that `node0` has 350k stake? That can't be right. 😁